File: src/day2.rs
use fancy_regex::Regex;
use crate::util::read_file;

pub fn run() {
    part_1();
    part_2();
}

fn part_2() {
    let lines = read_file("inputs/day2.txt");
    let re = Regex::new(r"(?<one>\d+)-(?<two>\d+) (?<char>\w): (?<pass>.*)").unwrap();
    let mut valid = 0;
    for line in lines.split('\n') {
        let groups = re.captures(&line).unwrap().unwrap();
        let one = groups.name(&"one").unwrap().as_str().parse::<usize>().unwrap() - 1;
        let two = groups.name(&"two").unwrap().as_str().parse::<usize>().unwrap() - 1;
        let tmp: Vec<char> = groups.name(&"char").unwrap().as_str().chars().collect();
        let ch = tmp.get(0).unwrap();
        let pass: Vec<char> = groups.name(&"pass").unwrap().as_str().chars().collect();
        if &pass[one] == ch {
            if &pass[two] != ch {
                valid += 1;
            }
        } else {
            if &pass[two] == ch {
                valid += 1;
            }
        }
    }
    println!("There are {} valid passwords", valid);
}

fn part_1() {
    let lines = read_file("inputs/day2.txt");
    let re = Regex::new(r"(?<min>\d+)-(?<max>\d+) (?<char>\w): (?<pass>.*)").unwrap();
    let mut valid = 0;
    for line in lines.split('\n') {
        let groups = re.captures(&line).unwrap().unwrap();
        let min = groups.name(&"min").unwrap().as_str().parse().unwrap();
        let max = groups.name(&"max").unwrap().as_str().parse().unwrap();
        let tmp: Vec<char> = groups.name(&"char").unwrap().as_str().chars().collect();
        let ch = tmp.get(0).unwrap();
        let pass = groups.name(&"pass").unwrap().as_str();
        let ch_count = pass.chars().filter(|c| c == ch).count();
        if ch_count >= min && ch_count <= max {
            valid += 1;
        }
    }
    println!("There are {} valid passwords", valid);
}

File: elixir/lib/day2.ex
defmodule Day2 do
    def run do
        lines = Input.readFile("day2")
        part1(lines)
        part2(lines)
    end

    defp deconstruct(line) do
        # Map.values returns them alphabetically by capture group
        ~r/(?<a>\d+)-(?<b>\d+) (?<c>\w): (?<d>.*)/
            |> Regex.named_captures(line)
            |> Map.values
    end

    defp charCountValid(passLine) do
        [min, max, char, pass] = deconstruct(passLine)
        [min, max] = [min, max] |> Enum.map(&String.to_integer/1)
        {:ok, char} = Regex.compile(char)
        case char |> Regex.scan(pass) |> Enum.count do
            n when min <= n and n <= max -> 1
            _ -> 0
        end
    end

    defp part1(lines) do
        IO.puts "2-1: There are [#{lines |> Enum.map(&charCountValid/1) |> Enum.sum}] valid passwords"
    end

    defp xor(a, b) do (a and !b) or (!a and b) end

    defp charLocValid(passLine) do
        [one, two, char, pass] = deconstruct(passLine)
        [one, two] 
            |> Enum.map(&(String.to_integer(&1) - 1))
            |> Enum.map(&(String.at(pass, &1) == char))
            |> Enum.reduce(false, &xor/2)
    end

    defp part2(lines) do
        IO.puts "2-2: There are [#{
            lines |> Enum.map(&charLocValid/1) |> Enum.filter(&(&1)) |> Enum.count
        }] valid passwords"
    end
end